File: algorithm/Swap-Nodes-in-Pairs.py
# Given a linked list, swap every two adjacent nodes and return its head.

# You may not modify the values in the list's nodes, only nodes itself may be changed.
# Example:
# Given 1->2->3->4, you should return the list as 2->1->4->3.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        h = head
        while h.next is not None and h.next.next:
            h.val, h.next.val = h.next.val, h.val
            h = h.next.next

File: algorithm/SwapNodesinPairs.go
package main

import "fmt"

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	a := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 3,
			},
		},
	}
	printListNode(swapPairs(a))
}
func printListNode(head *ListNode) {
	ans := make([]int, 0)
	for head != nil {
		ans = append(ans, head.Val)
		head = head.Next
	}
	fmt.Println(ans)
}

func swapPairs(head *ListNode) *ListNode {

	var helper func(*ListNode)

	helper = func(node *ListNode) {
		if node.Next == nil {
			return
		}
		if node.Next.Next == nil {
			return
		}
		temp := node.Next
		tempNext := temp.Next
		temp.Next = tempNext.Next
		tempNext.Next = temp
		node.Next = tempNext
		helper(node.Next.Next)
	}
	temp := &ListNode{
		Val:  0,
		Next: head,
	}
	helper(temp)
	return temp.Next
}
